src/main/kotlin/org/rust/lang/core/types/infer/TyLowering.kt

```diff
@@ -0,0 +1,338 @@
+            }
+
+            is RsFnPointerType -> {
+                val paramTypes = type.valueParameters.map { p -> p.typeReference?.let { lowerTy(it, null) } ?: TyUnknown }
+                TyFunction(paramTypes, type.retType?.let { it -> it.typeReference?.let { lowerTy(it, null) } ?: TyUnknown } ?: TyUnit.INSTANCE)
+            }
+
+            is RsTraitType -> {
+                var hasSizedUnbound = false
+                val traitBounds = type.polyboundList.mapNotNull {
+                    if (it.hasQ) {
+                        hasSizedUnbound = true
+                        null
+                    } else {
+                        val path = it.bound.traitRef?.path ?: return@mapNotNull null
+                        val res = rawMultiResolvePath(path).singleOrNull() ?: return@mapNotNull null
+                        instantiatePathGenerics(path, res.element, res.resolvedSubst, PathExprResolver.default, withAssoc = true)
+                            .downcast<RsTraitItem>()
+                    }
+                }
+                if (type.isImpl) {
+                    val sized = type.knownItems.Sized
+                    val traitBoundsWithImplicitSized = if (!hasSizedUnbound && sized != null) {
+                        traitBounds + BoundElement(sized)
+                    } else {
+                        traitBounds
+                    }
+                    return TyAnon(type, traitBoundsWithImplicitSized)
+                }
+                if (traitBounds.isEmpty()) return TyUnknown
+                val lifetimeBounds = type.polyboundList.mapNotNull { it.bound.lifetime }
+                val regionBound = lifetimeBounds.firstOrNull()?.resolve() ?: defaultTraitObjectRegion ?: ReStatic
+                TyTraitObject(traitBounds, regionBound)
+            }
+
+            is RsMacroType -> {
+                if (type.calculateMacroExpansionDepth() >= DEFAULT_RECURSION_LIMIT) return TyUnknown
+                val expansion = type.macroCall.expansion as? MacroExpansion.Type ?: return TyUnknown
+                lowerTy(expansion.type, null)
+            }
+
+            else -> TyUnknown
+        }
+    }
+
+    private fun rawMultiResolvePath(path: RsPath): List<RsPathResolveResult<RsElement>> {
+        return resolvedNestedPaths[path]
+            ?: path.reference?.rawMultiResolve()
+            ?: emptyList()
+    }
+
+    private fun <T : RsElement> instantiatePathGenerics(
+        path: RsPath,
+        element: T,
+        subst: Substitution,
+        resolver: PathExprResolver,
+        withAssoc: Boolean
+    ): BoundElement<T> {
+        if (element !is RsGenericDeclaration) return BoundElement(element, subst)
+
+        val genericParameters = genericParametersCache.getOrPut(element) { element.getGenericParameters() }
+        val psiSubstitution = pathPsiSubst(path, element, givenGenericParameters = genericParameters)
+
+        val typeSubst = psiSubstitution.typeSubst.entries.associate { (param, value) ->
```

Comment:
Why `RsPsiSubstitution.toSubst` was inlined?

Reply:
First of all, it gives shorter call stacks that is handy for profiling :D
![image](https://user-images.githubusercontent.com/3221931/189697356-8c3642a1-34f0-4211-94ee-6fee6be2e2e4.png)

Seriously, I think that will help with future optimization, since now the recursion is clearly visible (lowerTy -> instantiatePathGenerics -> lowerTy -> ...).

Also I think it's a better design, despite the function is now quite large. Previously we had a method `RsPsiSubstitution.toSubst` that ignored `RsPsiSubstitution.assoc` because, well, it returned `Substitution` that did not contain `assoc`. But in `instantiatePathGenerics` we immediately mapped `psiSubstitution.assoc` after calling `RsPsiSubstitution.toSubst`. So i find that division to `instantiatePathGenerics` and `toSubst` quite artificial and decided that inlining `toSubst` would be a better design